### PR TITLE
dependency graph: support the dependency between file and references

### PR DESCRIPTION
**Pull request by ansyral**
we can report uid dependency directly in plugins, which would be resolved at later phase.

### CODE REVIEW

**Comment by superyyrrzz** [comment]
> namespace Microsoft.DocAsCode.Build.Engine.Incrementals

copyright header

---
Refers to: src/Microsoft.DocAsCode.Build.Engine/Incrementals/DependencyItemSourceInfo.cs:1 in 783b9c3. [](commit_id = 783b9c33898cf2858fe05e5631e8cf17b71ef465, deletion_comment = False)

File: src/Microsoft.DocAsCode.Build.Engine/Incrementals/DependencyItemSourceInfo.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            return Comparer.GetHashCode(Value) ^ (SourceType.GetHashCode() >> 1);
+        }
+
+        public static implicit operator DependencyItemSourceInfo(string info)
+        {
+            return info == null ? null : new DependencyItemSourceInfo { Value = info, SourceType = DependencyItemSourceType.File };
+        }
+    }
+
+    public enum DependencyItemSourceType
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.Engine/Incrementals/DependencyItemSourceInfo.cs]: >DependencyItemSourceType [](start = 16, length = 24)

move to a new file

File: src/Microsoft.DocAsCode.Build.Engine/Incrementals/DependencyItem.cs [hunk 2]
```diff
@@ -5,35 +5,46 @@ namespace Microsoft.DocAsCode.Build.Engine.Incrementals
...
-        public DependencyItem(string from, string to, string reportedBy, string type)
+        public DependencyItem(DependencyItemSourceInfo from, DependencyItemSourceInfo to, DependencyItemSourceInfo reportedBy, string type)
         {
             From = from;
             To = to;
             ReportedBy = reportedBy;
             Type = type;
         }
 
+        public void UpdateFrom(DependencyItemSourceInfo f)
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/Incrementals/DependencyItem.cs]: >UpdateFrom [](start = 20, length = 10)

using public property `set` is enough?

File: src/Microsoft.DocAsCode.Build.Engine/Incrementals/DependencyItemSourceInfo.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+    using Newtonsoft.Json;
+
+    public class DependencyItemSourceInfo : IEquatable<DependencyItemSourceInfo>
+    {
+        private StringComparer Comparer
+        {
+            get
+            {
+                return SourceType == DependencyItemSourceType.File ? FilePathComparer.OSPlatformSensitiveStringComparer : StringComparer.Ordinal;
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/Incrementals/DependencyItemSourceInfo.cs]: =>

File: src/Microsoft.DocAsCode.Build.Engine/Incrementals/DependencyItemSourceInfo.cs (hunk 1 above)
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.Engine/Incrementals/DependencyItemSourceInfo.cs]: and move to plugin project?

---
In reply to: [102121624](https://github.com/dotnet/docfx/pull/1323#discussion_r102121624) [](ancestors = 102121624)

File: src/Microsoft.DocAsCode.Build.Engine/HostService.cs [hunk 4]
```diff
@@ -331,9 +353,37 @@ public void ReportDependencyFrom(FileModel currentFileModel, string from, string
...
+                string fromKey = fromT == DependencyItemSourceType.File ?
+                    IncrementalUtility.GetDependencyKey(currentFileModel.OriginalFileAndType.ChangeFile((RelativePath)currentFileModel.OriginalFileAndType.File + (RelativePath)from)) :
+                    from;
                 string toKey = IncrementalUtility.GetDependencyKey(currentFileModel.OriginalFileAndType);
-                ReportDependencyCore(fromKey, toKey, toKey, type);
+                ReportDependencyCore(new DependencyItemSourceInfo { SourceType = fromT, Value = fromKey }, toKey, toKey, type);
+            }
+        }
+
+        public void ReportReference(FileModel currentFileModel, string reference, string referenceType)
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.Engine/HostService.cs]: >string [](start = 82, length = 6)

DependencyItemSourceType?

File: src/Microsoft.DocAsCode.Build.Engine/HostService.cs [hunk 5]
```diff
@@ -331,9 +353,37 @@ public void ReportDependencyFrom(FileModel currentFileModel, string from, string
...
+            if (currentFileModel == null)
+            {
+                throw new ArgumentNullException(nameof(currentFileModel));
+            }
+            if (string.IsNullOrEmpty(reference))
+            {
+                throw new ArgumentNullException(nameof(reference));
+            }
+            DependencyItemSourceType rt;
+            if (referenceType == null || !Enum.TryParse<DependencyItemSourceType>(referenceType, out rt))
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.Engine/HostService.cs]: `referenceType` is "File" or "Reference"?

Comment (ansyral) [inline src/Microsoft.DocAsCode.Build.Engine/HostService.cs]: oops, only Reference, but need to tell different kinds of `Reference`

**Comment by superyyrrzz** [comment]
:shipit:

File: src/Microsoft.DocAsCode.Build.Engine/Incrementals/DependencyItem.cs (hunk 2 above)
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.Engine/Incrementals/DependencyItem.cs]: Or create new DependencyItem instance?

---
In reply to: [102122717](https://github.com/dotnet/docfx/pull/1323#discussion_r102122717) [](ancestors = 102122717)

File: src/Microsoft.DocAsCode.Build.Engine/Incrementals/DependencyItem.cs [hunk 6]
```diff
@@ -62,7 +73,7 @@ public override string ToString()
 
         public override int GetHashCode()
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.Engine/Incrementals/DependencyItem.cs]: Consider following case:
```cs
var set = new HashSet<DependencyItem>();
var item = new DependencyItem{ ... };
set.Add(item);
item.UpdateFrom(...);
Console.WriteLine(set.Contains(item)); // false
```

Comment (ansyral) [inline src/Microsoft.DocAsCode.Build.Engine/Incrementals/DependencyItem.cs]: nice catch! need to remove and then add the item to update the hashcode

**Comment by vwxyzh** [comment]
:shipit:

**Pull request merged**